test/rateLimiters/tokenBucket.test.ts

```diff
@@ -0,0 +1,265 @@
+import redis from 'redis-mock';
+import { RedisClientType } from 'redis';
+import TokenBucket from '../../src/rateLimiters/tokenBucket';
+
+const CAPACITY = 10;
+// FIXME: Changing the refill rate effects test outcomes.
+const REFILL_RATE = 1; // 1 token per second
+
+let limiter: TokenBucket;
+let client: RedisClientType;
+let timestamp: number;
+const user1 = '1';
+const user2 = '2';
+const user3 = '3';
+const user4 = '4';
+
+async function getBucketFromClient(
+    redisClient: RedisClientType,
+    uuid: string
+): Promise<RedisBucket> {
+    return redisClient.get(uuid).then((res) => JSON.parse(res || '{}'));
+}
+
+async function setTokenCountInClient(
+    redisClient: RedisClientType,
+    uuid: string,
+    tokens: number,
+    time: number
+) {
+    const value: RedisBucket = { tokens, timestamp: time };
+    await redisClient.set(uuid, JSON.stringify(value));
+}
+
+xdescribe('Test TokenBucket Rate Limiter', () => {
+    beforeEach(async () => {
+        // Initialize a new token bucket before each test
+        // create a mock user
+        // intialze the token bucket algorithm
+        client = redis.createClient();
+        limiter = new TokenBucket(CAPACITY, REFILL_RATE, client);
+        timestamp = new Date().valueOf();
+    });
+
+    describe('TokenBucket returns correct number of tokens and updates redis store as expected', () => {
+        test('after an ALLOWED request', async () => {
+            // Bucket intially full
+            const withdraw5 = 5;
+            expect(limiter.processRequest(user1, timestamp, withdraw5).tokens).toBe(
+                CAPACITY - withdraw5
+            );
+            const tokenCountFull = await getBucketFromClient(client, user1);
+            expect(tokenCountFull).toBe(CAPACITY - withdraw5);
+
+            // Bucket partially full but enough time has elapsed to fill the bucket since the last request and
+            // has leftover tokens after reqeust
+            const initial = 6;
+            const partialWithdraw = 1;
+            await setTokenCountInClient(client, user2, initial, timestamp);
+            expect(
+                limiter.processRequest(
+                    user2,
+                    timestamp + 1000 * (CAPACITY - initial),
+                    initial + partialWithdraw
+                )
+            ).toBe(CAPACITY - (initial + partialWithdraw));
```

Comment:
These case reads to me like:
line 58 - set token bucket to have 6 tokens
line 60 - process request withdrawing 7 tokens (6+1)
line 65 - expect the 3 tokens in the bucket (10 - (6 + 1)) when I think there will be -1 tokens in the bucket

I'm having a hard time following all this redis stuff so I might be wrong. Need more time.

Reply:
This also needs to be processRequest().tokens to access the tokens property on the returned object

Reply:
line 62 is saying that enough time has elapsed to refill the bucket since the last request (4000 ms). So when we withdraw 7 tokens there should be 3 remaining. Was trying to get after a case where the bucket fills enough between requests.